creator-node/src/blacklistManager.js

```diff
@@ -250,48 +288,188 @@ class BlacklistManager {
+            }
+          }
+        }
+
+        if (errors.length > 0) { throw new Error(errors.toString()) }
+        break
+      }
+      case REDIS_SET_INVALID_TRACKIDS_KEY:
+      case REDIS_SET_BLACKLIST_SEGMENTCID_KEY:
+      case REDIS_SET_BLACKLIST_TRACKID_KEY:
+      case REDIS_SET_BLACKLIST_USERID_KEY:
+      default: {
+        if (!data || data.length === 0) return
+        try {
+          const resp = await redis.sadd(redisKey, data)
+          this.logDebug(`redis set add ${redisKey} response ${resp}`)
+        } catch (e) {
+          throw new Error(`Unable to add ${redisKey}:${data}: ${e.toString()}`)
+        }
+        break
+      }
     }
   }
 
   /**
    * Removes key with value to redis. If value does not exist, redis should ignore.
-   * @param {string} key
-   * @param {number[]} value
+   * @param {string} redisKey type of value
+   * @param {number[] | string[] | Object} data either array of userIds, trackIds, CIDs, or <trackId: [CIDs]>
    */
-  static async removeFromRedis (key, value) {
-    if (!value || value.length === 0) return
+  static async removeFromRedis (redisKey, data) {
+    switch (redisKey) {
+      case REDIS_MAP_BLACKLIST_SEGMENTCID_TO_TRACKID_KEY: {
+        const errors = []
+        for (let [trackId, cids] of Object.entries(data)) {
+          trackId = parseInt(trackId)
+          for (const cid of cids) {
+            const redisCIDKey = this.getRedisBlacklistSegmentToTrackIdKey(cid)
+            try {
+              await redis.srem(redisCIDKey, trackId)
+            } catch (e) {
+              errors.push(`Unable to remove ${redisCIDKey}:${trackId}: ${e.toString()}`)
+            }
+          }
+        }
+
+        if (errors.length > 0) { throw new Error(errors.toString()) }
+        break
+      }
+      case REDIS_SET_BLACKLIST_SEGMENTCID_KEY:
+      case REDIS_SET_BLACKLIST_TRACKID_KEY:
+      case REDIS_SET_BLACKLIST_USERID_KEY:
+      default: {
+        if (!data || data.length === 0) return
+        try {
+          const resp = await redis.srem(redisKey, data)
+          this.logDebug(`redis set remove ${redisKey} response ${resp}`)
+        } catch (e) {
+          throw new Error(`Unable to remove ${redisKey}:${data}: ${e.toString()}`)
+        }
+        break
+      }
+    }
+  }
+
+  static async isServable (cid, trackId = null) {
     try {
-      const resp = await redis.srem(key, value)
-      logger.info(`redis set remove ${key} response: ${resp}`)
+    // If the CID is not in the blacklist, allow serve
+      const CIDIsInBlacklist = await this.CIDIsInBlacklist(cid)
+      if (!CIDIsInBlacklist) return true
+
+      // If the CID is in the blacklist and an invalid trackId was passed in, do not serve
+      // Also, if the CID is not of track type and is in the blacklist, do not serve anyway
+      if (!trackId || isNaN(trackId) || trackId < 0 || !Number.isInteger(trackId)) return false
```

Comment:
I actually think this is ok/we should keep this because this part of the fn is only reached if the cid is in the bl. this acts the same as what exists on master right now where if the cid is in the bl, do not serve. wdyt?

Reply:
Are you sure? If the caller doesn't pass a trackId through, this if statement will cause isServable to return false and the check in the calling function says if(!servable) error()

Reply:
i read the code badly. ignore